Add testing for isFlattened and IsFlattenedArray Unsafe methods

ehrenjulzert opened the pull request:
Part of testing for #13614
Depends on #13924 being merged

Signed-off-by: Ehren Julien-Neitzert <ehren.julien-neitzert@ibm.com>

hangshao0 commented:
> What should the priority of each test be?

The should be functional sanity tests.

> Should Point2D, FlattenedPoint2D, or TestClasses.java be renamed?

Probably you want to rename `FlattenedPoint2D` to be `valueTypePoint2D`. "Flattened" is more like a JVM behavior which can be controlled by command line option `-XX:ValueTypeFlatteningThreshold=`. You may want to rename `TestClasses` to something more meaningful.

> The tests currently only pass on a some of the testing variations (such as _ValueTypeTests_2) because other variations do not pass -XX:ValueTypeFlatteningThreshold=99999 and -XX:+EnableArrayFlattening to java

The test should be run with `-XX:ValueTypeFlatteningThreshold=99999 -XX:+EnableArrayFlattening`, which turns on flattening on value type classes and arrays.

ehrenjulzert commented:
> They should be functional sanity tests

Sorry, I wasn't clear about what I meant by priority. In [ValueTypeTests.java](https://github.com/eclipse-openj9/openj9/blob/master/test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java#L666) each test has an annotation with a specific priority defined, ranging from 1 to 6 (such as `@Test(priority=3)`). I'm not sure which of these 6 priorities the tests should be assigned.

> The test should be run with -XX:ValueTypeFlatteningThreshold=99999 -XX:+EnableArrayFlattening

when you run all the tests using `make _sanity` all the tests in ValueTypeTests.java are run 8 times. And each time they are run, different command line arguments are used (see [Valhalla/playlist.xml](https://github.com/eclipse-openj9/openj9/blob/master/test/functional/Valhalla/playlist.xml)). Out of the 8 times they're run, only variants 2, 3, 5, and 6 have the correct `-XX:ValueTypeFlatteningThreshold` and `-XX:+EnableArrayFlattening` arguments. I'm wondering if there's a way to specify that these tests should only be run with specific command line arguments, or if I will have to move them all to a new test file?

hangshao0 commented:
> ranging from 1 to 6 (such as @Test(priority=3)). I'm not sure which of these 6 priorities the tests should be assigned.

@llxia, do you if we are using theses priority numbers and if yes how do we decide the priority ?

> when you run all the tests using make _sanity all the tests in ValueTypeTests.java are run 8 times

That probably means you need to move your new test into a new file/test target.

ehrenjulzert commented:
I have renamed the tests and moved them to a new file so that `make _sanity` always passes. Now all that is left is to figure out if priority matters.

ehrenjulzert commented:
@llxia has let me know that I can ignore priority as the tests do not need to be run in a particular order

hangshao0 commented:
I see `WIP` is removed from the title. I assume this is ready to be reviewed ?

ehrenjulzert commented:
yep, it's ready for review

hangshao0 commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTestClasses.java:
```diff
@@ -0,0 +1,51 @@
+/*******************************************************************************
+ * Copyright (c) 2018, 2021 IBM Corp. and others
```
The start year of copyright should be 2021.

hangshao0 commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTestClasses.java:
```diff
@@ -0,0 +1,51 @@
+package org.openj9.test.lworld;
+
+
+public class ValueTypeUnsafeTestClasses {
```
I suggest to use `ValueTypeUnsafeTests` as the class name.

ehrenjulzert replied:
ValueTypeUnsafeTests is already the name of the class that stores the tests. Maybe the tests and classes should be merged into one file?

hangshao0 replied:
> ValueTypeUnsafeTests is already the name of the class that stores the tests

OK. Then leave it the way it is now.

hangshao0 commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java:
```diff
@@ -0,0 +1,96 @@
+/*******************************************************************************
+ * Copyright (c) 2018, 2021 IBM Corp. and others
```
The start year of copyright should be 2021.

hangshao0 commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java:
```diff
@@ -0,0 +1,96 @@
+                try {
+                        Field f = Unsafe.class.getDeclaredField("theUnsafe");
+                        f.setAccessible(true);
+                        myUnsafe = (Unsafe)f.get(null);
```
The public method `getUnsafe()` in unsafe class is not working ? It can be used to return the unsafe instance.
https://github.com/eclipse-openj9/openj9/blob/c37e61a9d5caa7d4a946f911fb62810e59e84030/jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java#L1073

ehrenjulzert replied:
Huh, it was not working when I was testing before but it's working now (maybe I was using sun.misc.Unsafe?). Regardless, I'll change it to getUnsafe() now

hangshao0 commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java:
```diff
@@ -0,0 +1,96 @@
+	}
+
+
+	@Test
```
Incorrect indent.

hangshao0 commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java:
```diff
@@ -0,0 +1,96 @@
+
+	@Test
+        static public void testFlattenedFieldIsFlattened() throws Throwable {
+                ValueTypePoint2D p = new ValueTypePoint2D(new ValueTypeInt(0), new ValueTypeInt(0));
```
`ValueTypeInt.i` is 0 by default. I suggest initializing using non-zero values.

hangshao0 commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java:
```diff
@@ -0,0 +1,96 @@
+
+        @Test
+        static public void testFlattenedArrayIsFlattened() throws Throwable {
+                ValueTypePoint2D[] pointAry = {};
```
You can also add an assert that `pointAry.getClass().isArray()` is true. Similarly in the `testRegularArrayIsNotFlattened()` below.

hangshao0 commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java:
```diff
@@ -0,0 +1,96 @@
+
+        @Test
+        static public void testRegularArrayIsNotFlattened() throws Throwable {
+                Integer[] ary = {};
```
Integer might be migrated to value type in the future. It is safer to use another non-value type class, e.g. `java.lang.Thread`.

hangshao0 commented on test/functional/Valhalla/testng.xml:
```diff
@@ -36,4 +35,9 @@
+                <classes>
+                        <class name="org.openj9.test.lworld.ValueTypeUnsafeTests"/>
+                </classes>
+        </test>
```
Incorrect indent.

hangshao0 commented:
I think the description https://github.com/eclipse-openj9/openj9/pull/13961#issue-1058943942 needs to be updated. At least the following 2 points should be removed:

> Should Point2D, FlattenedPoint2D, or TestClasses.java be renamed?
> The tests currently only pass on a some of the testing variations (such as _ValueTypeTests_2) because other variations do not pass -XX:ValueTypeFlatteningThreshold=99999 and -XX:+EnableArrayFlattening to java
> 

Also please mention #13614 in the description of this PR and in the commit message.

ehrenjulzert commented:
All unresolved issues should be addressed now in f4bdecc

llxia commented:
Jenkins test sanity xlinuxval jdknext

llxia reviewed (approved):
LGTM
